pkg/controller/endpointslice/endpointslice_tracker.go

```diff
@@ -19,58 +19,82 @@ package endpointslice
-// by the EndpointSlice controller.
+// endpointSliceTracker tracks EndpointSlices and their associated generation to
+// help determine if a change to an EndpointSlice has been processed by the
+// EndpointSlice controller.
 type endpointSliceTracker struct {
-	// lock protects resourceVersionsByService.
+	// lock protects generationsByService.
 	lock sync.Mutex
-	// resourceVersionsByService tracks the list of EndpointSlices and
-	// associated resource versions expected for a given Service.
-	resourceVersionsByService map[types.NamespacedName]endpointSliceResourceVersions
+	// generationsByService tracks the generations of EndpointSlices for each
+	// Service.
+	generationsByService map[types.NamespacedName]generationsBySlice
 }
 
 // newEndpointSliceTracker creates and initializes a new endpointSliceTracker.
 func newEndpointSliceTracker() *endpointSliceTracker {
 	return &endpointSliceTracker{
-		resourceVersionsByService: map[types.NamespacedName]endpointSliceResourceVersions{},
+		generationsByService: map[types.NamespacedName]generationsBySlice{},
 	}
 }
 
-// Has returns true if the endpointSliceTracker has a resource version for the
+// Has returns true if the endpointSliceTracker has a generation for the
 // provided EndpointSlice.
 func (est *endpointSliceTracker) Has(endpointSlice *discovery.EndpointSlice) bool {
 	est.lock.Lock()
 	defer est.lock.Unlock()
 
-	rrv, ok := est.relatedResourceVersions(endpointSlice)
+	gfs, ok := est.generationsForSlice(endpointSlice)
 	if !ok {
 		return false
 	}
-	_, ok = rrv[endpointSlice.Name]
+	_, ok = gfs[endpointSlice.UID]
 	return ok
 }
 
-// Stale returns true if this endpointSliceTracker does not have a resource
-// version for the provided EndpointSlice or it does not match the resource
-// version of the provided EndpointSlice.
+// Stale returns true if this endpointSliceTracker does not have a generation
+// for the provided EndpointSlice or it is greater than the generation of the
+// provided EndpointSlice.
 func (est *endpointSliceTracker) Stale(endpointSlice *discovery.EndpointSlice) bool {
 	est.lock.Lock()
 	defer est.lock.Unlock()
 
-	rrv, ok := est.relatedResourceVersions(endpointSlice)
+	gfs, ok := est.generationsForSlice(endpointSlice)
 	if !ok {
 		return true
 	}
-	return rrv[endpointSlice.Name] != endpointSlice.ResourceVersion
+	g, ok := gfs[endpointSlice.UID]
+	if !ok {
+		return true
+	}
+	return g > endpointSlice.Generation
+}
+
+// OneOrMoreStale returns true if this endpointSliceTracker does not have a
+// generation for one or more of the provided EndpointSlices or one or more of
+// the provided EndpointSlices have older generations than the corresponding
+// tracked ones.
+func (est *endpointSliceTracker) OneOrMoreStale(service *v1.Service, endpointSlices []*discovery.EndpointSlice) bool {
+	nn := types.NamespacedName{Name: service.Name, Namespace: service.Namespace}
+	gfs, ok := est.generationsByService[nn]
+	if !ok {
+		return false
```

Comment:
Why we return true here and false in Sta() method in the corresponding place?

If there is no entry for a service it means we aren't aware of staleness of this object, so this false seems fine, but I don't understand why we return true in line 72.

[I guess I'm missing something subtle, but can you please add a comment there to explain...]

Reply:
Yeah this was very confusing and I think that was mostly due to the naming of the function. I've updated this function to use `HasNewerSlices` as the name and updated the logic a bit too match the name. I think that's more reflective of what I'm trying to accomplish here. Unfortunately it means that we may end up syncing with an informer cache that thinks we have EndpointSlices that the controller has already deleted. 

If we only synced when the informer cache only had the slices that the tracker was expecting, we could be waiting forever, especially if a user created their own slice and the tracker didn't expect it to exist.

Reply:
I've updated this to also track EndpointSlices that the controller expects to be deleted so the last part of my comment above is no longer true.